fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreUniqueIndexTest.java

```diff
@@ -303,74 +338,458 @@ public void changeIndexAtFixedSubspaceKey() throws Exception {
+        }
+
+        private void assertNoUniquenessViolations(final FDBRecordContext context)
+                throws InterruptedException, ExecutionException {
+            assertThat(recordStore.scanUniquenessViolations(nonUniqueIndex).asList().get(), empty());
+            assertThat(context.ensureActive().getRange(recordStore.indexUniquenessViolationsSubspace(nonUniqueIndex)
+                            .range()).asList().get(),
+                    Matchers.hasSize(0));
+        }
+    }
+
+    /**
+     * Factory for {@link NeverAllowClearUniquenessViolations}.
+     */
+    @AutoService(IndexMaintainerFactory.class)
+    public static class NeverAllowClearUniquenessViolationsFactory implements IndexMaintainerFactory {
+
+        @Nonnull
+        @Override
+        public Iterable<String> getIndexTypes() {
+            return Collections.singletonList(NO_UNIQUE_CLEAR_INDEX_TYPE);
+        }
+
+        @Nonnull
+        @Override
+        public IndexValidator getIndexValidator(Index index) {
+            return new IndexValidator(index);
+        }
+
+        @Nonnull
+        @Override
+        public IndexMaintainer getIndexMaintainer(@Nonnull IndexMaintainerState state) {
+            return new NeverAllowClearUniquenessViolations(state);
+        }
+    }
+
+    private static class NeverAllowClearUniquenessViolations extends IndexMaintainer {
+        IndexMaintainer underlying;
+
+        public NeverAllowClearUniquenessViolations(final IndexMaintainerState state) {
+            super(state);
+            underlying = new ValueIndexMaintainer(state);
+        }
+
+        @Nonnull
+        @Override
+        public RecordCursor<IndexEntry> scan(@Nonnull final IndexScanType scanType, @Nonnull final TupleRange range, @Nullable final byte[] continuation, @Nonnull final ScanProperties scanProperties) {
+            return underlying.scan(scanType, range, continuation, scanProperties);
+        }
+
+        @Nonnull
+        @Override
+        public <M extends Message> CompletableFuture<Void> update(@Nullable final FDBIndexableRecord<M> oldRecord, @Nullable final FDBIndexableRecord<M> newRecord) {
+            return underlying.update(oldRecord,  newRecord);
+        }
+
+        @Nonnull
+        @Override
+        public <M extends Message> CompletableFuture<Void> updateWhileWriteOnly(@Nullable final FDBIndexableRecord<M> oldRecord, @Nullable final FDBIndexableRecord<M> newRecord) {
+            return underlying.updateWhileWriteOnly(oldRecord, newRecord);
+        }
+
+        @Nonnull
+        @Override
+        public RecordCursor<IndexEntry> scanUniquenessViolations(@Nonnull final TupleRange range, @Nullable final byte[] continuation, @Nonnull final ScanProperties scanProperties) {
+            return underlying.scanUniquenessViolations(range, continuation, scanProperties);
+        }
+
+        @Override
+        public CompletableFuture<Void> clearUniquenessViolations() {
+            return super.clearUniquenessViolations();
```

Comment:
It might be worth highlighting that this method is the only one that _doesn't_ delegate to the value index maintainer, resulting in `clearUniquenessViolations` being a no-op. I think it's pretty clear form the index name that that's the intention, but highlighting that (somewhat subtle) difference is probably good for future maintainers trying to understand how it achieves that goal.

Reply:
Good call. I added a javadoc. And since this is now using the default implementation, instead of defining its own method to throw, I renamed it.